File: apps/agent/src/core/chat/chat_service.py
import prisma.enums
from langchain_core.messages import BaseMessage

from .utils import ChatUtils
from ..llm.llm import llm

class ChatService:
    async def get_chat(self):
        return await ChatUtils.get_messages()

    async def chat(self, message: str) -> str:
        history = await self.get_chat()
        response = await llm.get_completion(message, history)
        await ChatUtils.save_message(message, prisma.enums.Role.user)

        handled_response = self.handle_llm_response(response)
        await ChatUtils.save_message(handled_response, prisma.enums.Role.assistant)
        return handled_response

    def handle_llm_response(self, response: BaseMessage) -> str:
        if response.tool_calls:
            return self.handle_tool_call(response)
        return response.content

    def handle_tool_call(self, response: BaseMessage) -> str:
        message = f"Tool call: {response.tool_calls[0]['name']} with args: {response.tool_calls[0]['args']}"
        return message

File: apps/api/src/chat/chat.service.ts
import { Injectable, InternalServerErrorException } from '@nestjs/common';
import { CreateChatDto } from './dto/create-chat.dto';
import { UpdateChatDto } from './dto/update-chat.dto';
import { LlmService } from 'src/llm/llm.service';
import { PrismaService } from 'src/prisma/prisma.service';
import { Role } from '@prisma/client';

@Injectable()
export class ChatService {
  constructor(private readonly llmService: LlmService, private readonly prismaService: PrismaService) {}
  async create(createChatDto: CreateChatDto): Promise<string> {
    try{
      const messages = await this.prismaService.message.findMany({})
  
      const newMessage = await this.prismaService.message.create({
        data: {
          content: createChatDto.user_query,
          role: Role.user
        }
      })
  
      const response = await this.llmService.chat([...messages, newMessage])
      Promise.all(
        response.map(r => 
          this.prismaService.message.create({
            data: {
              content: r.content,
              role: r.role
            }
          })
        )
      ).catch(error => {
        console.error("Failed to save assistant messages:", error);
      });
  
      return response.find(message => message.role === Role.assistant)?.content || "No response from assistant"
    } catch(error) {
      if (error instanceof Error) {
        throw new Error("Error in ChatService.create: " + error.message)
      }
      throw new InternalServerErrorException("Error in ChatService.create: " + error)
    }
  }

  async get() {
    try{
      const messages = await this.prismaService.message.findMany({
        where: {
          role: {
            not: Role.tool
          }
        }
      })
      return messages
    } catch(error) {
      throw new InternalServerErrorException("Error in ChatService.getChat: " + error)
    }
  }

  async clear() {
    try{
      await this.prismaService.message.deleteMany({})
    } catch(error) {
      throw new InternalServerErrorException("Error in ChatService.clearChat: " + error)
    }
  }
}
